Objects/listobject.c

```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
         lo.values = saved_ob_item;
     }
 
+
+    /* The pre-sort check: here's where we decide which compare function to use.
+     * How much optimization is safe? We test for homogeneity with respect to 
+     * several properties that are expensive to check at compare-time, and 
+     * set ms appropriately. */
+    if (saved_ob_size > 1) {
```

Comment:
For `<= 1`, I just didn't want to mess with anything. For `2,3`, I just didn't want to have to write additional code to set `ms` appropriately. I think it makes more sense to just have one block of code, instead of a separate block at the end that sets `ms->key_compare = safe_object_compare` for `== 2,3`. My benchmarking indicates that the gain would be negligble, and not worth an extra 3-4 lines...